systems/controllers/osc/operational_space_control.h

```diff
@@ -0,0 +1,238 @@
+  drake::systems::EventStatus DiscreteVariableUpdate(
+      const drake::systems::Context<double>& context,
+      drake::systems::DiscreteValues<double>* discrete_state) const;
+
+  // Output function
+  void CalcOptimalInput(const drake::systems::Context<double>& context,
+                        systems::TimestampedVector<double>* control) const;
+
+  // Input/Output ports
+  int state_port_;
+  int fsm_port_;
+
+  // Discrete update
+  int prev_fsm_state_idx_;
+  int prev_event_time_idx_;
+
+  // Map position/velocity from model with spring to without spring
+  Eigen::MatrixXd map_position_from_spring_to_no_spring_;
+  Eigen::MatrixXd map_velocity_from_spring_to_no_spring_;
+
+  // Map from trajectory names to input port indices
+  std::map<std::string, int> traj_name_to_port_index_map_;
+
+  // RBT's.
+  const RigidBodyTree<double>& tree_w_spr_;
+  const RigidBodyTree<double>& tree_wo_spr_;
+
+  // Size of position, velocity and input of the RBT without spring
+  int n_q_;
+  int n_v_;
+  int n_u_;
+
+  // Size of holonomic constraint and total contact constraints
+  int n_h_;
+  int n_c_;
+
+  // robot input limits
+  Eigen::VectorXd u_min_;
+  Eigen::VectorXd u_max_;
+
+  // flag indicating whehter using osc with finite state machine or not
+  bool used_with_finite_state_machine_;
+
+  // flag indicating whether to print the tracking related values or not
+  bool print_tracking_info_;
+
+  // floating base model flag
+  bool is_quaternion_;
+
+  // MathematicalProgram
+  std::unique_ptr<drake::solvers::MathematicalProgram> prog_;
+  // Decision variables
+  drake::solvers::VectorXDecisionVariable dv_;
+  drake::solvers::VectorXDecisionVariable u_;
+  drake::solvers::VectorXDecisionVariable lambda_c_;
+  drake::solvers::VectorXDecisionVariable lambda_h_;
+  drake::solvers::VectorXDecisionVariable epsilon_;
+  // Cost and constraints
+  drake::solvers::LinearEqualityConstraint* dynamics_constraint_;
+  drake::solvers::LinearEqualityConstraint* holonomic_constraint_;
+  drake::solvers::LinearEqualityConstraint* contact_constraints_;
+  std::vector<drake::solvers::LinearConstraint*> friction_constraints_;
+  std::vector<drake::solvers::QuadraticCost*> tracking_cost_;
+
+  // OSC cost members
+  /// Using u cost would push the robot away from the fixed point, so the user
+  /// could consider usnig acceleration cost instead.
+  Eigen::MatrixXd W_input_;  // Input cost weight
+  Eigen::MatrixXd W_joint_accel_;  // Joint acceleration cost weight
+
+  // OSC constraint members
+  bool with_input_constraints_ = true;
+
+  // (flat ground) Contact constraints and friction cone cnostraints
+  std::vector<int> body_index_ = {};
+  std::vector<Eigen::VectorXd> pt_on_body_ = {};
+  double mu_ = -1;  // Friction coefficients
+  double w_soft_constraint_ = -1;
+
+  // `state_` is the finite state machine state when the contact constraint is
+  // active. If `state_` is empty, then the constraint is always active.
+  std::vector<int> state_;
```

Comment:
`state_` probably deserves a better name.`fsm_state_when_active_` is one suggestion, but you can probably find something better!